File: erlang/pe009.erl
-module(pe009).
-export([start/0]).

sum() -> 1000.

start() ->
  pe009(1, 2).

pe009(I, J) ->
  Bool = high_eq_lowmid(I, J),
  LastJ = sum() - (I + 1),
  if
    Bool ->
      K = sum() - (I + J),
      P = I * J * K,
      io:format("~w~n", [P]);
    J =:= LastJ ->
      pe009(I + 1, I + 2);
    true ->
      pe009(I, J + 1)
  end.

high_eq_lowmid(I, J) ->
   K = sum() - (I + J),
   I2 = I * I,
   J2 = J * J,
   K2 = K * K,
  [Low , Mid, High] = lists:sort([I2, J2, K2]),
  (Low + Mid =:= High).

File: ocaml/pe009.ml
let isum = 1000 ;;

let high_eq_lowmid i j  = 
  let k  = isum - (i + j) in
  let i2 = i * i in
  let j2 = j * j in
  let k2 = k * k in
  (i2 > j2 && i2 > k2 && i2 = j2 + k2) ||
  (j2 > i2 && j2 > k2 && j2 = i2 + k2) ||
  (k2 > i2 && k2 > j2 && k2 = i2 + j2) ;;

let rec pe009 i j =
  let lastj = isum - (i + 1) in
  if high_eq_lowmid i j then
    let k = isum - (i + j) in
    Printf.printf "%d\n" (i * j * k)
  else
    if j = lastj then
      pe009 (i + 1) (i + 2)
    else
      pe009 i (j + 1) ;;

let main () =
  pe009 1 2 ;;

let () = main ()
